[RLlib] Replay Buffer Docs (#24683)
ray-project/ray: doc/source/rllib/rllib-replay-buffers.rst

> .. _replay-buffer-api-reference-docs:

Reviewer: Wait, this is the "reference-docs" and the one above is the "references-api-docs", no?

##############
Replay Buffers
##############

Quick Intro to Replay Buffers in RL
=====================================

When we talk about replay buffers in reinforcement learning, we generally mean a buffer that stores and replays experiences collected from interactions of our agent(s) with the environment.

> In python, a simple buffer can be implemented by a list to which elements are added and sampled from.

Reviewer: ```suggestion
In python, a simple buffer can be implemented by a list to which elements are added and later sampled from.
```

Such buffers are used mostly in off-policy learning algorithms. This makes sense intuitively because these algorithms can learn from

> experiences that are stored in the buffer but where produced by a previous version of the policy (or even a completely different policy).

Reviewer: ```suggestion
experiences that are stored in the buffer, but where produced by a previous version of the policy (or even a completely different "behavior policy").
```

Sampling Strategy
-----------------

> When sampling from a replay buffer, we choose which experiences to train our agent with. A straightforward strategy to choose these
> samples that has proven effective for many algorithms is uniform sampling. That is, choosing a random item from the buffer each time we sample.
> A strategy that has proven better in many cases is `Prioritized Experiences Replay (PER) <https://arxiv.org/abs/1511.05952>`__.
> In PER, single items in the buffer are assigned a priority which denotes their significance, or in simple terms, how much we expect
> to learn from these items. As a consequence, experiences with a higher priority are made more likely to be sampled.

Reviewer: ```suggestion
When sampling from a replay buffer, we choose which experiences to train our agent with. A straightforward strategy that has proven effective for many algorithms is to pick these
samples uniformly at random.
A more advanced strategy (proven better in many cases) is `Prioritized Experiences Replay (PER) <https://arxiv.org/abs/1511.05952>`__.
In PER, single items in the buffer are assigned a (scalar) priority value, which denotes their significance, or in simpler terms, how much we expect
to learn from these items. Experiences with a higher priority are more likely to be sampled.
```

Eviction Strategy
-----------------

> A buffer is naturally limited in its capacity to hold experiences. In the course of running an algorith, a buffer will eventually reach

Reviewer: typo: "algorithm"

its capacity and in order to make room for new experiences, we need to delete (evict) older ones. This is generally done on a first-in-first-out basis.
For your algorithms this means that buffers with a high capacity give the opportunity to learn from older samples, while smaller buffers
make the learning process more on-policy. An exception from this strategy is made in buffers that implement `reservoir sampling <https://www.cs.umd.edu/~samir/498/vitter.pdf>`__.

Replay Buffers in RLlib
=======================

> RLlib comes with a set of extendable replay buffers built in.

Reviewer: Add: All buffers support the 2 basic methods `add` and `sample`.

.. dropdown:: **Executable example script**
    :animate: fade-in-slide-down

    .. literalinclude:: ../../../rllib/examples/replay_buffer_api.py
        :emphasize-lines: 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77
        :language: python
        :start-after: __sphinx_doc_replay_buffer_api_example_script_begin__
        :end-before: __sphinx_doc_replay_buffer_api_example_script_end__

.. tip:: Because of its prevalence, most Q-learning algorithms support PER. The priority update step that is needed is embedded into their training iteration functions.
.. warning:: If your custom buffer requires extra interaction, you will have to change the training iteration function, too!

Specifying a buffer type works the same way as specifying an exploration type.
Here are three ways of specifying a type:

.. dropdown:: **Changing a replay buffer configuration**
    :animate: fade-in-slide-down

    .. literalinclude:: ../../../rllib/examples/documentation/replay_buffer_demo.py
        :language: python
        :start-after: __sphinx_doc_replay_buffer_type_specification__begin__
        :end-before: __sphinx_doc_replay_buffer_type_specification__end__

> Apart from specifying a type, other parameters in that config are:

Reviewer: Maybe mention here that all the other args (non-type) are basically c'tor kwargs of the used type.

I also think we should mention `capacity` first, as it's the most important one. Then maybe storage unit, and only then the more exotic ones like worker_side_prio, etc..

Author: But that would not be correct! There are non-ctor args in there as well. Like worker_side_prioritization.
Capacity is currently the first example of a c'tor arg.

I've changed the this part a bit - lmk if this still needs improvement :)

#. Parameters that define how algorithms interact with replay buffers.
    e.g. ``worker_side_prioritization`` to decide where to compute priorities

#. Constructor arguments to instantiate the replay buffer.
    e.g. ``capacity`` to limit the buffer's size

#. Call arguments for underlying replay buffer methods.
    e.g. ``prioritized_replay_beta`` is used by the :py:class:`~ray.rllib.utils.replay_buffers.multi_agent_prioritized_replay_buffer.MultiAgentPrioritizedReplayBuffer` to call the ``sample()`` method of every underlying :py:class:`~ray.rllib.utils.replay_buffers.prioritized_replay_buffer.PrioritizedReplayBuffer`

.. tip:: Most of the time, only 1. and 2. are of interest. 3. is an advanced feature that supports use cases where a :py:class:`~ray.rllib.utils.replay_buffers.multi_agent_replay_buffer.MultiAgentReplayBuffer` instantiates underlying buffers that need constructor or default call arguments.

ReplayBuffer Base Class
-----------------------

> The base :py:class:`~ray.rllib.utils.replay_buffers.replay_buffer.ReplayBuffer` class only supports storing and replaying

Reviewer: Mention again here the 2 basic methods that "everyone must know" :)
add and sample

experiences in different :py:class:`~ray.rllib.utils.replay_buffers.replay_buffer.StorageUnit`\s.
Advanced buffer types add functionality while trying to retain compatibility through inheritance.
The following is an example of the most basic scheme of interaction with a :py:class:`~ray.rllib.utils.replay_buffers.replay_buffer.ReplayBuffer`.

.. literalinclude:: ../../../rllib/examples/documentation/replay_buffer_demo.py
    :language: python
    :start-after: __sphinx_doc_replay_buffer_basic_interaction__begin__
    :end-before: __sphinx_doc_replay_buffer_basic_interaction__end__

Building your own ReplayBuffer
------------------------------

> Here is an example of how to implement your own toy example of a ReplayBuffer class and make SimpleQ use it:

Reviewer: Cool!

.. literalinclude:: ../../../rllib/examples/documentation/replay_buffer_demo.py
    :language: python
    :start-after: __sphinx_doc_replay_buffer_own_buffer__begin__
    :end-before: __sphinx_doc_replay_buffer_own_buffer__end__

For a full implementation, you should consider other methods like ``get_state()`` and ``set_state()``.

> A more extensive example is `our implementation <https://github.com/ray-project/ray/blob/master/rllib/utils/replay_buffers/reservoir_buffer.py>`__ of `reservoir sampling <https://www.cs.umd.edu/~samir/498/vitter.pdf>`__, the :py:class:`~ray.rllib.utils.replay_buffers.reservoir_replay_buffer.ReservoirBuffer`.

Reviewer: Didn't we rename it to `ReservoirReplayBuffer`?
